src/utils/messaging-outbox/src/services/outbox_transactional_processor.rs

```diff
@@ -320,35 +321,80 @@ impl ProducerRelayJob {
                 )
                 .await?;
 
+            // Collecting failing consumers as we go
+            let mut failing_consumer_names: HashSet<String> = HashSet::new();
+
             // Feed consumers if they are behind this message
             // We must respect the sequential order of messages,
             // but individual consumers may process each message concurrently
             for message in unprocessed_messages {
                 // Prepare consumer invocation tasks
                 let mut consumer_tasks = Vec::new();
                 for consumer_name in &self.consumer_names {
+                    // Skip consumers, which are failing
+                    if failing_consumer_names.contains(consumer_name) {
+                        continue;
+                    }
+
+                    // Skip consumers, which are already beyond this message ID
                     let boundary_id = consumption_boundaries
                         .get(consumer_name)
                         .copied()
                         .unwrap_or_else(|| OutboxMessageID::new(0));
                     if boundary_id < message.message_id {
+                        // Non-failing consumer which hasn't seen this message is a task to execute
                         consumer_tasks.push((consumer_name.as_str(), &message));
                     }
                 }
 
-                // Consume concurrently
-                // TODO: error recovery
-                use futures::{StreamExt, TryStreamExt};
-                futures::stream::iter(consumer_tasks)
-                    .map(Ok)
-                    .try_for_each_concurrent(
-                        /* limit */ None,
-                        |(consumer_name, message)| async move {
-                            self.invoke_consumer(consumer_name, message).await?;
-                            Ok(())
-                        },
-                    )
-                    .await?;
+                #[derive(Error, Debug)]
+                #[error("Outbox message consumption error")]
+                struct OutboxMessageConsumptionError {
```

Comment:
Why make this struct nested? This will prevent you from even testing this error.

Reply:
It is only used locally, yes

Reply:
Simplified it further, I don't really need `thiserror::Error` here, just a struct to be more explicit than tuple.